src/curve/constant-product/ConstantProductPair.sol

```diff
@@ -21,7 +21,23 @@ contract ConstantProductPair is ReservoirPair {
     uint224 public kLast; // reserve0 * reserve1, as of immediately after the most recent liquidity event
 
     constructor(address aToken0, address aToken1) Pair(aToken0, aToken1)
-    {} // solhint-disable-line no-empty-blocks
+    {
+        swapFee = uint256(factory.get(keccak256("CP::swapFee")));
+
+        require(swapFee <= MAX_SWAP_FEE, "CP: INVALID_SWAP_FEE");
```

Comment:
I think you can DRY this with either an abstract constant or abstract function. Like `abstract constant PAIR_SWAP_FEE_NAME: string`

Reply:
okay so I can't explicitly define an `abstract constant`. The closest thing is declaring it in `IPair.sol` and then defining it in ConstantProduct/Stable. 

So we got 2 options
1. define a `PAIR_SWAP_FEE_NAME` in `IPair.sol`. Then we still have to call `this.PAIR_SWAP_FEE_NAME()` in `updateSwapFee`. Cuz it's not defined in `Pair.sol` but in the individual curves. 
2. we define a `string internal swapFeeName` in `Pair.sol`, and then set that at the constructor of each curve. I tried making this immutable. We cannot even if it is `bytes32` because this variable will be set in the individual curves, not the abstract contract itself. Immutable variables can only be set in the constructor of the defining contract. 
   - this is the current impl

Reply:
2 is better, avoid CALL. Solidity sucks for not letting you use abstract consts in the contract that declares them... What about if you define it as a constructor argument for Pair.sol so that the child is forced to declare its curve name to the parent when it constructs the parent?

Then it can be immutable

Reply:
Passing the `swapFeeName` as a parameter into the base class is a genius idea. 

Except that for solidity right now it can't have non-value types as immutables. Meaning we can't have `string` immutables, only `bytes32`. So strings can be constants but can't be immutables. 

So anyway, to make this work, we have to store `swapFeeName` as a bytes32, which is:

```solidity
swapFeeName = keccak256(abi.encodePacked(aPairSwapFeeName); 
```

So this would differ from the other constant types, which are string. And then we have to use `.get()` for this case instead of `.read()`. Both of which I can live with.

Reply:
Works for me. I have a feeling during gas golfing we may move away from strings to bytes32 strings or something anyway. So let's leave it for now. Can resolve